### ISSUE TITLE
[Feature request] RMSE and MAE using bins

### ISSUE THREAD

**Issue by Expertium** [open-spaced-repetition/fsrs-vs-sm17#35]
Currently, it's very hard to compare results [here](https://github.com/open-spaced-repetition/fsrs-vs-sm17/blob/main/README.md) with the results from [the benchmark](https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Benchmark).
I recommend the following:
1) Rename the "RMSE" column to "RMSE (raw)"
2) Add two new columns: "RMSE (bins)" and "MAE (bins)"
3) Add a text note: "RMSE (bins) and MAE (bins) are added to allow a comparison between these values and the values on the benchmark page of FSRS wiki"

Also, while it may be too late for that, it would be better to have one repo for all "FSRS vs SuperMemo" comparisons, instead of 2 different repos.

**Comment by L-M-Sherlock** [comment]
> 3\. RMSE (bins) and MAE (bins) are added to allow a comparison between these values and the values on the benchmark page of FSRS wiki

How to comparison them when the datasets are different?

**Comment by L-M-Sherlock** [comment]
> 2\. RMSE (bins)

Done in:
- #36

I decide to remove MAE from other evaluation, because MAE is not a good metric. For more about the metric, please see: https://github.com/open-spaced-repetition/spaced-repetition-algorithm-metric

**Comment by L-M-Sherlock** [comment]
> First of all, the benchmark page on the wiki and the "FSRS vs SuperMemo" repos must use **exactly** the same metrics. This may involve making an automated benchmarking setup, which is something that @L-M-Sherlock didn't do for the benchmark, and I had to test everything manually.

I plan to open a new repo to store the data and write a script to automate it.

> Second of all (less important, but still), the same number of significant digits must be displayed.

The two comparisons both use 4 significant digits.

**Comment by Expertium** [comment]
Please add a text note that explains why there are 2 different values of RMSE.

As for significant digits, fsrs-vs-sm15 and fsrs-vs-sm17 have the same number of significant digits, but the benchmark page on the wiki has s different number of significant digits.

**Comment by L-M-Sherlock** [comment]
https://github.com/open-spaced-repetition/fsrs-vs-sm15/commit/f2259ec9fcec8cb300ad7a85305497ffb3557256 added RMSE(bins)

**Comment by Expertium** [comment]
While looking att he code, I noticed that you are calculating Cohen's d, although it's not displayed anywhere on the page. I have something better for you: https://sci-hub.se/10.3102/10769986025002101
This "common language effect size" (CLES for short, or _Vargha-Delaney A_) has 3 advantages over Cohen's d:
1) It works for any distribution of your data, not just normal distribution
2) It works for discrete data as well, not just for continuous data
3) Most importantly, it is far easier to interpret

Here's the python code with an example:

```
FSRS_rmse = [0.05, 0.02, 0.04, 0.07, 0.01]
SM_rmse = [0.06, 0.02, 0.05, 0.07, 0.015]

def CLES(x, y):
    n = len(x)
    m = len(y)
    # step 1: combine two lists without sorting the new list
    combined = list(x) + list(y)
    # step 2: compute ranks of all observations
    combined_rank = list(scipy.stats.rankdata(combined))
    # step 3: compute the sum of ranks for elements of list x (the first n elements of the combined list)
    R_x = np.sum(combined_rank[:n])
    # step 4: compute the final statistic using this magical formula
    A_x = (R_x / n - (n + 1) / 2) / m
    return A_x

print(f"Common language effect size={100 * CLES(SM_rmse, FSRS_rmse):.1f}%")
```
This outputs a number between 0 and 1 that can be interpreted as "the probability that a randomly chosen FSRS user will have a lower value of RMSE than a randomly chosen SuperMemo user". Note that if you change `CLES(SM_rmse, FSRS_rmse)` to `CLES(FSRS_rmse, SM_rmse)`, you will get one minus that probability. In other words, you will get "the probability that a randomly chosen SuperMemo user will have a lower value of RMSE than a randomly chosen FSRS user"

**Comment by L-M-Sherlock** [comment]
Thanks, I will consider it in the subsequent evaluation.

By the way, I have built the dataset for comparison: https://github.com/open-spaced-repetition/fsrs-benchmark

**Comment by Expertium** [comment]
Btw, using the benchmark data from the wiki (59 datapoints, RMSE), the common language effect size of FSRS v4 vs FSRS v3 is 69.3%, which is quite large. This means that if we compare a random user of FSRS v4 to a random user of FSRS v3, in 69.3% of all cases RMSE will be lower in v4.
The problem with this value is that we are comparing a _random_ user to _another random_ user. If the same user switches from v3 to v4, the probability that RMSE will be lower for that one user is way higher. So maybe we shouldn't use this after all. Especially since this involves using another metric, making if more confusing. The users will be like "So, this is the probability of something-something blah-blah-blah, which, in turn, is based on some other number called RESM or whatever?".

**Comment by L-M-Sherlock** [comment]
Fine. I will save the raw metrics data at first.

![image](https://github.com/open-spaced-repetition/fsrs-vs-sm17/assets/32575846/6874d0ab-1223-410d-b036-3abe8cc5f929)

**Comment by L-M-Sherlock** [comment]
Which algorithms should we compared in this benchmark repo?

**Comment by Expertium** [comment]
Algorithms:
1) FSRS v4 online
2) FSRS v4 offline
3) FSRS v3
4) SM-15
5) SM-17

It's important to compare offline and online versions of FSRS. It's also important to compare them to the previous version.

Metrics:
Log-loss
RMSE (raw)
RMSE (bins)

**Comment by L-M-Sherlock** [comment]
I don't want to compare FSRS v4 online here, because it's too slow. And I can't compare SM-15/SM-17, because SuperMemo doesn't provide API to predict R from revlogs.

**Comment by Expertium** [comment]
Then v4, v3 and an LSTM.

**Comment by Expertium** [comment]
Right now my biggest concern is that comparing the values in this repo and on the wiki is very difficult. So if users ask "Ok, FSRS is great compared to an LSTM and it's own previous version, but what about SuperMemo?", answering that question will be difficult. Plus, it's better to keep everything benchmark-related in one repo, and you are kind of doing the exact opposite.

**Comment by L-M-Sherlock** [comment]
> (although I would remove it)

But Woz used it in his comparison:

> Feb 22, 2018: We have finally added a simulation of Algorithm SM-2 to SuperMemo for Windows and have come up with the average least squares metric of 53.5685% (for Algorithm SM-2). For comparison, [Algorithm SM-17](https://supermemopedia.com/wiki/Algorithm_SM-17) results in 37.1202% (a million repetitions dataset).

source: http://supermemopedia.com/wiki/Algorithm_SM-17_vs._older_SuperMemos

**Comment by Expertium** [comment]
Yes, but it's not as informative and not as easy to interpret as RMSE based on bins.

**Comment by L-M-Sherlock** [comment]
> you are kind of doing the exact opposite.

If I implement them in one repo, the development will be delayed to the next year. Because I don't like to develop a huge project. I prefer to solve them one by one.

**Comment by L-M-Sherlock** [comment]
The current result:

Total number of users: 71

Total review logs for evaluation: 4,632,965

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| FSRS v4 | 0.3879 | 0.3346 | 0.0453 |

https://github.com/open-spaced-repetition/fsrs-benchmark

**Comment by Expertium** [comment]
I would change the wording to "Total number of reviews:"

**Comment by L-M-Sherlock** [comment]
Total number of users: 71

Total number of reviews for evaluation: 4,632,965

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| FSRS v4 | 0.3879 | 0.3346 | 0.0453 |
| LSTM | 0.4199 | 0.3425 | 0.0662 |
| FSRS v3 | 0.4890 | 0.3633 | 0.1204 |

**Comment by Expertium** [comment]
RMSE of FSRS v3 seems to be a lot higher than before
![image](https://github.com/open-spaced-repetition/fsrs-vs-sm17/assets/83031600/4db07bf2-2206-4c3d-8b55-5e550fdbcc2d)
Btw, do you use weighting by the number of reviews?

**Comment by L-M-Sherlock** [comment]
Yeah, I weighed the metrics by the number of reviews.

**Comment by Expertium** [comment]
> Yeah, I weighed the metrics by the number of reviews. 

I would suggest ln(n_reviews) instead, just like in the original benchmark.

> And I also applied time series splits in the benchmark for avoiding the information leak. 

But that matters only for online versions, no?

**Comment by L-M-Sherlock** [comment]
For a cautious comparison, it's necessary. Because not algorithm can know the review from the future. For practical purposes, it also holds. We use the parameters trained in the past reviews to predict our memory state in the future.

**Comment by user1823** [comment]
> For practical purposes, it also holds. We use the parameters trained in the past reviews to predict our memory state in the future.

I don't get it. When the final parameters are obtained, the parameters have already been trained on all the reviews. The evaluation begins after this. So, how does time series split help make the evaluation fairer?

By the way, if does make the evaluation fairer, do you use time series split in the release version of the FSRS optimizer?

**Comment by L-M-Sherlock** [comment]
```python
    tscv = TimeSeriesSplit(n_splits=n_splits)
    dataset.sort_values(by=["review_time"], inplace=True)
    for train_index, test_index in tscv.split(dataset):
        train_set = dataset.iloc[train_index].copy()
        test_set = dataset.iloc[test_index].copy()
        optimizer.S0_dataset_group = (
            train_set[train_set["i"] == 2]
            .groupby(by=["r_history", "delta_t"], group_keys=False)
            .agg({"y": ["mean", "count"]})
            .reset_index()
        )
        optimizer.define_model()
        _ = optimizer.pretrain(dataset=train_set, verbose=verbose)
        trainer = Trainer(
            train_set,
            test_set,
            optimizer.init_w,
            n_epoch=n_epoch,
            lr=lr,
            batch_size=batch_size,
        )
        w_list.append(trainer.train(verbose=verbose))
        testsets.append(test_set)

    p, y = predict(w_list, testsets)
```

The weights are optimized in the trainset, and used to predict the testset.

**Comment by L-M-Sherlock** [comment]
> I would suggest ln(n_reviews) instead, just like in the original benchmark.

OK. I will apply it to fsrs-vs-sm15 and fsrs-vs-sm17 for consistency.

**Comment by L-M-Sherlock** [comment]
### Weighted by number of reviews

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| FSRS v4 | 0.3879 | 0.3346 | 0.0453 |
| LSTM | 0.4199 | 0.3425 | 0.0662 |
| FSRS v3 | 0.4890 | 0.3633 | 0.1204 |

### Weighted by ln(number of reviews)

| Algorithm | Log Loss | RMSE | RMSE(bins) |
| --- | --- | --- | --- |
| FSRS v4 | 0.3830 | 0.3311 | 0.0543 |
| FSRS v3 | 0.5132 | 0.3670 | 0.1326 |
| LSTM | 0.5788 | 0.3752 | 0.1385 |

**Comment by Expertium** [comment]
Just keep one weighting system. Having two tables makes it confusing.

**Comment by L-M-Sherlock** [comment]
Two points that I think would induce the higher RMSE and log loss for FSRS v3 and LSTM:

1. The new benchmark has more collections than the old one.
2. The new benchmark uses time series split.

**Comment by Expertium** [comment]
> Two tables would provide more solid evidences.

Not really. The old benchmark uses ln(number of reviews), so comparing the old values with new values only makes sense if the weighting system is the same.

**Comment by L-M-Sherlock** [comment]
What if the reader hasn't seen the old one? And I think it has made no sense to compare with the old benchmark, because many conditions have changed.

**Comment by L-M-Sherlock** [comment]
The method of splitting bins also changed. I used a new method to split the bins in calibration graph.

**Comment by Expertium** [comment]
> What if the reader hasn't seen the old one?

Then you still should only keep one weighting system to prevent confusion.

**Comment by L-M-Sherlock** [comment]
I don't think it would induce confusion. In academic papers, there are more weighting system and more conditions.

**Comment by Expertium** [comment]
Also, while conditions and data have changed, surely they didn't change _that_ much? I mean, RMSE has increased by 2x for FSRS v3, from 7% to 13%. Perhaps we need a way to filter out the outliers.

**Comment by L-M-Sherlock** [comment]
Here is the raw data:
https://github.com/open-spaced-repetition/fsrs-benchmark/tree/main/result

**Comment by Expertium** [comment]
No, I mean the metrics. Ideally, I want a .csv or .tsv file with and logloss, RMSE and RMSE (bins) for each collection.

**Comment by user1823** [comment]
@L-M-Sherlock, what do you think about
- adding Universal Metric to https://github.com/open-spaced-repetition/fsrs-benchmark
- adding [CLES](https://github.com/open-spaced-repetition/fsrs-vs-sm17/issues/35#issuecomment-1707791970) to each comparison (Benchmark, SM15, SM17)
- adding Memrise and SM-2 to https://github.com/open-spaced-repetition/fsrs-benchmark
- replacing the link to https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Benchmark on the Wiki page with the link to https://github.com/open-spaced-repetition/fsrs-benchmark?

> No, I mean the metrics. Ideally, I want a .csv or .tsv file with logloss, RMSE and RMSE (bins) for each collection.

@Expertium, the linked folder contains the metrics for each collection. The only problem is that these are not compiled so you have a separate json file for each collection and each algorithm (i.e., 71 × 3 = 213 files).

**Comment by Expertium** [comment]
> @L-M-Sherlock, what do you think about
> 
> * adding Universal Metric to https://github.com/open-spaced-repetition/fsrs-benchmark
> * adding [CLES](https://github.com/open-spaced-repetition/fsrs-vs-sm17/issues/35#issuecomment-1707791970) to each comparison (Benchmark, SM15, SM17)
> * adding Memrise and SM-2 to https://github.com/open-spaced-repetition/fsrs-benchmark
> * replacing the link to https://github.com/open-spaced-repetition/fsrs4anki/wiki/The-Benchmark on the Wiki page with the link to https://github.com/open-spaced-repetition/fsrs-benchmark?

I'm not Sherlock (obviously, lol), but I do want to comment
1) The universal metric is hard to understand, let's use RMSE for clarity and transparency
2) CLES has to be computed using some numbers. I don't like the idea of a metric that, in turn, is based on another metric, like a metric of a metric. It makes it harder to interpret, and I want to use simple metrics.
3) Many people have complained that because SM-2 and Memrise don't predict R natively, any attempt to compare them to FSRS/SuperMemo is meaningless.
4) Idk about the link

**Comment by L-M-Sherlock** [comment]
> No, I mean the metrics. Ideally, I want a .csv or .tsv file with logloss, RMSE and RMSE (bins) for each collection.

A JSON file contains the metrics for one collection:

<img width="568" alt="image" src="https://github.com/open-spaced-repetition/fsrs-vs-sm17/assets/32575846/68702928-eb04-4d96-abcf-ef44e927dca2">

If you still need .csv file, I will create one tomorrow.

**Comment by Expertium** [comment]
> If you still need .csv file, I will create one tomorrow.

Well, I don't want to manually go through all of them, and I can't code a script to do that for me, so I would appreciate your help.

**Comment by L-M-Sherlock** [comment]
[FSRSv3.csv](https://github.com/open-spaced-repetition/fsrs-vs-sm17/files/12543703/FSRSv3.csv)
[FSRSv4.csv](https://github.com/open-spaced-repetition/fsrs-vs-sm17/files/12543706/FSRSv4.csv)
[LSTM.csv](https://github.com/open-spaced-repetition/fsrs-vs-sm17/files/12543707/LSTM.csv)
